k2/csrc/rnnt_decode.cu

```diff
@@ -885,12 +952,36 @@ void RnntDecodingStreams::FormatOutput(const std::vector<int32_t> &num_frames,
           }
           arc.score = arc_info.score;
         }
-        out_map_data[oarc_idx0123] = arc_info.graph_arc_idx01;
-        arcs_out_data[oarc_idx0123] = arc;
+        out_map_data[oarc_idx01234] = arc_info.graph_arc_idx01;
+        arcs_out_data[oarc_idx01234] = arc;
+        if (arc_map_b != nullptr) {
+          t = t - num_padded_frames_data[oarc_idx0];
+          int32_t t2s2c_stream_idx0x = t2s2c_shape_row_splits1[t];
+          // oarc_idx0 is for stream.
+          int32_t t2s2c_stream_idx01 = t2s2c_stream_idx0x + oarc_idx0;
+          int32_t t2s2c_context_idx01x =
+                    t2s2c_shape_row_splits2[t2s2c_stream_idx01];
+          // oarc_idx2 is for context.
+          int32_t oarc_idx2 = oarc_idx012 - oarc_idx01x;
+          int32_t t2s2c_context_idx012 = t2s2c_context_idx01x + oarc_idx2;
+          // Manually map arc to super_final state(i.e. arc.label == -1)
+          // to blank_id == 0.
+          int32_t arc_label = 0;
+          if (arc.label != -1) {
+            arc_label = arc.label;
+          }
```

Comment:
I think you can set the `arc_map_token` to `-1` for final arcs,  and `k2.index_select` can map `-1` to default value  automatically.